README.md
# lista-clojure

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar lista-clojure-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2022 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-10-15
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-10-15
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/lista-clojure/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/lista-clojure/compare/0.1.0...0.1.1


test/lista_clojure/core_test.clj
(ns lista-clojure.core-test
  (:require [clojure.test :refer :all]
            [lista-clojure.core :refer :all]))

(deftest median
  (testing "Function media exists."
    (is (= (media 5) 5.0)))
  (testing "Accepts n arguments"
    (is (= (media 5) 5.0)))
    (is (= (media 5 4 ) 4.5))
    (is (= (media 5 4 3 2) 3.5))
  )


(deftest prod
  (testing "Function produto exists."
    (is (= (produto [5] [2]) 10)))
  (testing "Accepts n arguments"
    (is (= (produto [5] [1]) 5)))
    (is (= (produto [1 2] [4 5]) 14))
    (is (= (produto [1 2 3] [4 5 6]) 32)))

(deftest fibo
  (testing "Function get-fibo exists."
    (is (= (get-fibo 5) [5 3 1 1 0])))
  (testing "Testing different outcomes"
    (is (= (get-fibo 20) '(35 33 31 29 27 25 23 21 19 17 15 13 11 9 7 5 3 1 1 0))))
  (is (= (get-fibo 15) '(25 23 21 19 17 15 13 11 9 7 5 3 1 1 0)))
  (is (= (get-fibo 3) '(1 1 0))))

;; (deftest Fizz
;;   (testing "Function FizzBuzz exists."
;;     (is (= FizzBuzz ))))

(deftest rev
  (testing "Function de-reversa exists."
    (is (= (de-reversa "palavra") [nil \a \r \v \a \l \a \p])))
  (testing "Accepts vectors or strings"
  (is (= (de-reversa "pedra") [nil \a \r \d \e \p]))
  (is (= (de-reversa [1 2 3 4 5]) [nil 5 4 3 2 1]))))


(deftest tres
  (testing "Function tres-mais-freq exists."
    (is (= (tres-mais-freq ["palavra" "dois" "tres" "palavra" "dois" "palavra"]) '(["palavra" 3] ["dois" 2] ["tres" 1]))))
  (testing "Accepts numbers and strings"
    (is (= (tres-mais-freq ["palavra" "dois" "tres" "palavra" "dois" "palavra"]) '(["palavra" 3] ["dois" 2] ["tres" 1]))))
  (is (= (tres-mais-freq [1 2 3 4 4 3 5 2 4]) [[4 3] [2 2] [3 2]])))

(deftest size
  (testing "Function size-by exists."
    (is (= (size-by ["palavra" "dois" "tres" "aviao" "pato" "miscelaneo"]) '("miscelaneo" "palavra" "aviao"))))
  )

(deftest total-livros
  (testing "Function get-total exists."
    (is (= (get-total Livros) 144035))))

(deftest total-livros-sci-fi
  (testing "Function get-total-sci-fy exists."
    (is (= (get-total-sci-fy db) 53802))))

(deftest info-livros
  (testing "Function info exists."
    (is (= (info Livros) {:romance 32755, :sci-fi 49424, :horror 18623, :fantasy 43233}))))

(deftest eleven
  (testing "Function onze exists."
    (is (= (onze 5 6 10) [50 40 30 20 10]))))


(deftest gold
  (testing "Function goldbach exists."
    (is (= (goldbach 6) [5 1] )))
  (testing "Testing different outcomes"
    (is (= (goldbach 56) [53 3]))
    (is (= (goldbach 200) [199 1]))
    (is (= (goldbach 1000) [997 3]))
    )
  )


src/lista_clojure/core.clj
(ns lista-clojure.core
  (:gen-class))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Hello, World!"))

;; ----------------------------------------------------

;; 1
(defn media [& args]
  (float (/ (apply + args) (count args))))

;; ----------------------------------------------------

;; 2

(defn produto [a b]
  (loop [indice 0 saida 0]
    (if (= indice (count (map * a b)))
      saida
      (recur (inc indice) (+ saida (nth (map * a b) indice))))))

;; ----------------------------------------------------

;; 3

(defn fibonacci [n]
  (if (= n 1) 1
      (if (= n 0) 0
          (+
           (- n 1)
           (- n 2)))))

(fibonacci 3)

(defn get-fibo [n]
  (loop [indice 0 saida '()]
    (if (>= indice n)
      saida
      (recur (inc indice) (conj saida (fibonacci indice))))))


;; ----------------------------------------------------

;; 4

(defn FizzBuzz []
  (for [n (range 1 100)]
    (if (and (= 0 (mod n 3)) (= 0 (mod n 5)))
      "FizzBuzz"
      (if (= 0 (mod n 3))
        "Fizz"
        (if (= 0 (mod n 5))
          "Buzz"
          n)))))

(FizzBuzz)

;; ----------------------------------------------------

;; 5

(defn de-reversa [sequen]
  (loop [indice (count sequen) saida '[]]
    (if (< indice 0)
      saida
      (recur (dec indice) (conj saida (get sequen indice))))))

;; ----------------------------------------------------

;; 6

(defn tres-mais-freq [col]
  (take 3
        (sort-by val >
                 (frequencies col))))

;; ----------------------------------------------------

;; 7

(defn size-by [col]
  (take 3
        (sort-by count > col)))

(size-by ["palavra" "dois" "tres" "aviao" "pato" "miscelaneo"])

;; ----------------------------------------------------

;; 8

(def Livros
  [{:title "Pride and Prejudice" :author "Austen" :genre :romance :copies-sold 15400}
   {:title "Fundation" :author "Isaac Asimov" :genre :sci-fi :copies-sold 28769}
   {:title "World War Z" :author "Brooks" :genre :horror :copies-sold 7250}
   {:title "Jurassic Park" :author "Crichton" :genre :sci-fi :copies-sold 13677}
   {:title "The Lord of Rings" :author "Tokien" :genre :fantasy :copies-sold 23450}
   {:title "Game of Thrones" :author "Martin" :genre :fantasy :copies-sold 19783}
   {:title "1984" :author "Orwell" :genre :sci-fi :copies-sold 2600}
   {:title "Dom Quixote" :author "Miguel Cervantes" :genre :romance :copies-sold 17355}
   {:title "Frankenstein" :author "Mary Shelley" :genre :horror :copies-sold 11373}
   {:title "Let Petit Prince" :author "Antoine de Saint-Exupéry" :genre :sci-fi :copies-sold
    4378}])

(map :genre Livros)

(defn get-total [col]
  (loop [index 0 total 0]
    (if (= index (count col))
      total
      (recur (inc index) (+ total (:copies-sold (get col index) ))))))

(get-total Livros)

;; ----------------------------------------------------

;; 9

(def db {:db
         {:nome "Biblioteca"
          :dados Livros
          :temp {:title "Dune" :author "Frank Herbert" :genre :sci-fi :copies-sold 4378}}})

(defn get-total-sci-fy [db]
  (loop [index 0 total (get (:temp (:db db)) :copies-sold)]
    (if (= index (count (:dados (:db db))))
      total
      (if (not (= (get (get (:dados (:db db)) index) :genre) :sci-fi))
        (recur (inc index) total)
        (recur (inc index) (+ total (:copies-sold (get (:dados (:db db)) index))))))))

(get-total-sci-fy db)

;; ----------------------------------------------------

;; 10

(defn info [col]
  (conj {:romance (apply + (map :copies-sold (filter #(= (:genre %) :romance) col)))
   :sci-fi (apply + (map :copies-sold (filter #(= (:genre %) :sci-fi) col)))
   :horror (apply + (map :copies-sold (filter #(= (:genre %) :horror) col)))
   :fantasy (apply + (map :copies-sold (filter #(= (:genre %) :fantasy) col)))})
  )

(info Livros)

;; Possibilidade de uma condicional para cada genero com um loop para gerar o map

(defn info2 [col]
  (loop [indice 0 saida '{:romance :copies-sold :sci-fi :copies-sold :horror :copies-sold :fantasy :copies-sold}]
    (if (= indice (count col))
      saida
      (cond
           (= (:genre (get col indice)) :romance) (recur (inc indice) (+ (:romance saida) (:copies-sold (get (:dados (:db db)) indice))))
           (= (:genre (get col indice)) :sci-fi) (recur (inc indice) (+ (:sci-fi saida) (:copies-sold (get (:dados (:db db)) indice))))
           (= (:genre (get col indice)) :horror) (recur (inc indice) (+ (:horror saida) (:copies-sold (get (:dados (:db db)) indice))))
           (= (:genre (get col indice)) :fantasy) (recur (inc indice) (+ (:fantasy saida) (:copies-sold (get (:dados (:db db)) indice))))
           :else "F")) 
      )
   )
 
(info2 Livros)

;; ----------------------------------------------------

;; 11

(defn onze [n x p]
  (loop [indice (+ 1 x) saida '()]
    (if (= (count saida) n)
      saida
      (if (= (mod indice p) 0)
        (recur (inc indice)(conj saida indice))
        (recur (inc indice) saida)
        )
      ))
  )

(onze 5 6 3)

;; ----------------------------------------------------

;; 12

;; 12) Conjectura de Goldbach: A conjectura de Goldbach diz que todo número par positivo
;; maior que 2 é a soma de dois números primos. Exemplo: 28 = 5 + 23. Foi confirmado
;; numericamente até números muito grandes. Escreva uma função para encontrar os
;; dois números primos que somam um dado inteiro par menos que 1.000.

(defn is-prime? [num]
  (empty?
   (filter
    #(= % 0)
    (map #(mod num %) (range 2 num)))))

(defn goldbach [n]
  (if (or (neg? n)(odd? n))
    "Numero impar ou negativo"
    (conj '[] (first (reverse (filter is-prime? (range 1 n)))) (- n (first (reverse (filter is-prime? (range 1 n))))))
    )
  )
  
(goldbach 1000)